maven/lib/dependabot/maven/file_parser/maven_dependency_parser.rb

```diff
@@ -0,0 +1,135 @@
+              # but mvn CLI will not generate dependency tree for it unless we start from the parent.
+              # In that case we can just skip it and focus only on current file and it's sub-modules.
+              unless File.exist?(output_file)
+                Dependabot.logger.warn("Dependency tree output file not found: #{output_file}")
+                next
+              end
+
+              dependency_tree = JSON.parse(File.read(output_file))
+              extract_dependencies_from_tree(pom, dependency_set, dependency_tree)
+            end
+          end
+
+          dependency_set
+        end
+
+        private
+
+        sig do
+          params(pom: Dependabot::DependencyFile,
+                 dependency_set: Dependabot::FileParsers::Base::DependencySet,
+                 dependency_tree: T::Hash[String, T.untyped]).void
+        end
+        def extract_dependencies_from_tree(pom, dependency_set, dependency_tree)
+          traverse_tree = T.let(-> {}, T.proc.params(node: T::Hash[String, T.untyped]).void)
+          traverse_tree = lambda do |node|
+            artifact_id = node["artifactId"]
+            group_id = node["groupId"]
+            version = node["version"]
+            type = node["type"]
+            classifier = node["classifier"].to_s.empty? ? nil : node["classifier"]
+            scope = node["scope"]
+
+            groups = scope == "test" ? ["test"] : []
+            dependency_set << Dependabot::Dependency.new(
+              name: "#{group_id}:#{artifact_id}",
+              version: version,
+              package_manager: "maven",
+              requirements: [{
+                requirement: version,
+                file: nil,
+                groups: groups,
+                source: nil,
+                metadata: {
+                  packaging_type: type,
+                  classifier: classifier,
+                  pom_file: pom.name
+                }
+              }]
+            )
+
+            node["children"]&.each { |child| traverse_tree.call(child) }
+          end
+
+          traverse_tree.call(dependency_tree)
+        end
+
+        sig do
+          params(dependency_files: T::Array[Dependabot::DependencyFile], temp_path: String)
+            .returns(String)
+        end
+        def create_directory_structure(dependency_files, temp_path)
+          # Find the topmost directory level by finding the minimum number of "../" sequences
+          relative_top_depth = 0
+          dependency_files.each do |pom|
+            # This logic might generate incorrect depth if relative path is more complex,
+            # e.g. "../subdir/../pom.xml", but it should guarantee that resulted depth will be higher
+            # that actual maximul depth, which means that we will not escape the temporary directory.
+            # This is sufficient for our use case.
+            depth = pom.name.scan("../").length
```

Comment:
You might be able to use `Pathname.new(p).relative_path_from(directory).to_path` to generate a normalized relative path before counting the navigation.